### ISSUE TITLE
[css-values] Add a way to set longhands to the corresponding expansion of a shorthand value

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#8055]
### Rationale

There are some ways to set longhands to values that can't be represented at specified-value time, so they just serialize as empty string. In particular:

 - [Pending-substitution values](https://drafts.csswg.org/css-variables-1/#pending-substitution-value):
    ```js
    document.body.style.margin = "var(--m)";
    document.body.style.marginTop; // "" per spec
    ```
 - Expanded `toggle()` values (see #6764)
    ```js
    document.body.style.margin = "toggle(1px, 2px)";
    document.body.style.marginTop; // "" per #6764
    ```
 - Expanded [system fonts](https://drafts.csswg.org/css-fonts-4/#font-prop)
    ```js
    document.body.style.font = "small-caption";
    document.body.style.fontFamily; // "" in Gecko, Blink, WebKit
    ```
 - Values set by a legacy shorthand with broader grammar (see #7195)
    ```js
    document.body.style.webkitBackgroundClip = "text";
    document.body.style.backgroundClip; // "" in Backgrounds L3 (presumably, see #7195)
    ```
    ```js
    document.body.style.webkitBorderImage = "100% / 1px";
    document.body.style.borderImageWidth; // "" in WebKit, since 1px forces border-width when set via -webkit-border-image
    ```

This doesn't seem ideal, since we lose round-tripping (assigning empty string in CSSOM will just remove the declaration) and it's hard to tell what the value is. All these cases have in common that the value is set from some shorthand, so `cssText` can help serializing such shorthand, but this produces poor results when some longhand has been modified afterwards (see #2515):

```js
document.body.style.cssText = "margin: var(--m); margin-top: 1px";
document.body.style.marginLeft; // "" :( and the following don't help:
document.body.style.cssText; // "margin-right: ; margin-bottom: ; margin-left: ; margin-top: 1px;"
document.body.style.margin; // ""
```

### Proposal

So I had an idea to address all of this: introduce `from-shorthand( <custom-ident> ':' <whole-value>)`:

 - `from-shorthand()` is a [`<whole-value>`](https://drafts.csswg.org/css-values-5/#whole-value).
 - It's invalid if the `<custom-ident>` doesn't match the name of some shorthand ASCII case-insensitively.
 - When used on a longhand, it's invalid if the referred shorthand is not a shorthand of the longhand.
   When used on a shorthand, it's invalid if the referred shorthand is not a "super-shorthand" (i.e. it's invalid if there is a property which is a longhand of the current shorthand but is not a longhand of the referred shorthand).
 - It's invalid if the `<whole-value>` is an invalid value for the referred shorthand.
 - Custom properties don't have shorthands, so when used on an unregistered prop it should just be tokenized instead of resolving there.
 - At computed-value time, it basically resolves to the value that the current property would get if the referred shorthand had been set to the `<whole-value>`.
 - I'm using a `:` since it looks clearer, like a declaration. Function arguments typically use `,`  but that could get confusing in case the `<whole-value>` has commas. (Edited: initially suggested `;`)

Then:

```js
document.body.style.margin = "var(--m)";
document.body.style.marginTop; // "from-shorthand(margin: var(--m))"
```
```js
document.body.style.margin = "toggle(1px, 2px)";
document.body.style.marginTop; // "from-shorthand(margin: toggle(1px, 2px))"
```
```js
document.body.style.font = "small-caption";
document.body.style.fontFamily; // "from-shorthand(font: small-caption)"
```
```js
document.body.style.webkitBackgroundClip = "text";
document.body.style.backgroundClip; // "from-shorthand(-webkit-background-clip: text)" in L3
```
```js
document.body.style.webkitBorderImage = "100% / 1px";
document.body.style.borderImageWidth; // "from-shorthand(-webkit-border-image: 100% / 1px)" in WebKit
                                      // (the 100% is discarded and may be something else)
```
```js
document.body.style.cssText = "margin: var(--m); margin-top: 1px";
document.body.style.marginLeft; // "from-shorthand(margin: var(--m))"
document.body.style.cssText; // "margin-right: from-shorthand(margin: var(--m)); margin-bottom: from-shorthand(margin: var(--m)); margin-left: from-shorthand(margin: var(--m)); margin-top: 1px;"
document.body.style.margin; // ""
```

I guess there may be compat issues in the `var()` case, but hopefully people are not relying on getting an empty string.

**Comment by Loirooriol** [comment]
> but still doesn't let you reserialize the shorthand when one of the longhands is modified.

Yes, but it's not strange that a shorthand may not be able to represent all possible values of the longhands, and it's not such a big deal, since `cssText` can serialize longhands if shorthands don't work, and iterating a `CSSStyleDeclaration` goes through the longhands. So the important thing is being able to serialize the longhands.

Being able to specify which longhand to extract would offer more flexibility but it still doesn't guarantee that all shorthands will be serializable, and it opens the can of worms of using values from one property on another, which is difficult to define since some browsers normalize values at parse time by adding/removing/reordering components, which may affect whether it's still valid in the other property. Even in the same property it seems tricky if `from-shorthand()` can be mixed with other things, that's why I think it should be the only component and for the same property.

> using `:` instead would be much better imo

Yes I thought about `:` after filing the issue, may be the most clear option. Comma seems potentially confusing in e.g. `transition-property: from-shorthand(transition, margin, padding)` if you are not familiar with the function.

**Comment by Loirooriol** [comment]
I have edited the top post to use `:`. Also, if in the future we can avoid problems when assigning one expanded longhand to another longhand, the syntax could be extended to `from-shorthand(shorthand: value; longhand)`.

**Comment by Loirooriol** [comment]
Something that I missed is that we usually turn longhands into shorthands, so this should be allowed in shorthands as long as the referenced shorthand covers all the longhands (it's a supershorthand).

So for example, in Backgrounds L3,

```css
background-position: from-shorthand(background: 10px 20px red);
```
would behave as

```css
background-position: 10px 20px;
```

And in L4 it becomes a shorthand so it would expand to

```css
background-position-x: from-shorthand(background: 10px 20px red);
background-position-y: from-shorthand(background: 10px 20px red);
```
and behave as

```css
background-position-x: 10px;
background-position-y: 20px;
```

**Comment by tabatkins** [comment]
Yeah, so far I avoided exposing this just for lack of use-case, but if the lack is confusing authors, we should def expose it. I haven't given a lot of thought to what syntax to use; we've got a few here in the thread.

**Comment by kbabbitt** [comment]
In cases where there are multiple shorthands that include a longhand, how do we determine which shorthand we use when serializing out a `from-shorthand()` value?

```js
document.body.style.border = "2px dotted red";
console.log(document.style.borderLeftWidth);
```

Do I get:
1. `from-shorthand(border: 2px dotted red)`
2. `from-shorthand(border-left: 2px dotted red)`
3. `from-shorthand(border-width: 2px)`

Does the answer change if I set `document.body.style.borderLeft` instead?

**Comment by Loirooriol** [comment]
No, that should continue being `2px` for compat.

Values would only serialize with `from-shorthand()` when they wouldn't be serializable otherwise. And maybe if you specify `from-shorthand()` manually.

**Comment by kbabbitt** [comment]
Right, sorry, that was a badly chosen example. The main thing I was concerned about was making sure whatever we end up doing doesn't create any new requirements to store unparsed shorthands.

In the pending-substitution case, we already have to store the unparsed shorthand, so I suppose `from-shorthand()` would just use that.

`toggle()` I'd guess would end up similar if we ever figure out how to do it.

For system fonts, Blink stores "pending system font values" for each of the subproperties, which refer to the system font id. So that gives us enough to reconstruct what was specified.

Legacy shorthands look to me like a case where Blink would have to store something that we aren't today. But if there's just 1 or 2 of those, I'm not too bothered by that.

So on further thought I suppose it's not too much of an issue.

**Comment by Loirooriol** [comment]
Yes, I think in most cases the information is already known, and otherwise it's probably fine to let UAs come up with a valid answer.

For example, in `-webkit-border-image = 99% / 1px`, WebKit sets `border-image-width` to `1px 1px 1px 1px` with a special bool flag. So:
- This doesn't track the shorthand, but we know it must be `-webkit-border-image` because there is no other way to set that flag
- This doesn't track the original value either, but I think it's probably fine to e.g. serialize as `from-shorthand(-webkit-border-image: 100% / 1px)`, using the initial `border-image-slice: 100%` instead of the specified `99%`. Or as `from-shorthand(-webkit-border-image: none / 1px)`.

Or otherwise, if we want this to always track the original value, it also seems fine because it's only extra storage in corner cases.
